internal/exports/aws_s3_writer.go

```diff
@@ -0,0 +1,106 @@
+package exports
+
+import (
+	"bytes"
+	"fmt"
+	"image/png"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/innosat-mats/rac-extract-payload/internal/aez"
+	"github.com/innosat-mats/rac-extract-payload/internal/common"
+)
+
+const awsBucket = "mats-l0-artifacts"
+const awsS3Region = "eu-north-1"
+
+// AWSUpload uploads file content to target bucket
+func AWSUpload(uploader *s3manager.Uploader, key string, body io.Reader) {
+	_, err := uploader.Upload(&s3manager.UploadInput{
+		Bucket: aws.String(awsBucket),
+		Key:    aws.String(strings.ReplaceAll(key, "\\", "/")),
+		Body:   body,
+	})
+	if err != nil {
+		log.Printf("Failed to upload file %v, %v", key, err)
+	}
+
+}
+
+// AWSUploadFunc is the signature of an AWS upload function
+type AWSUploadFunc func(uploader *s3manager.Uploader, key string, body io.Reader)
+
+// AWSS3CallbackFactory generates callbacks for writing to S3 instead of disk
+func AWSS3CallbackFactory(
+	upload AWSUploadFunc,
+	project string,
+	awsDescriptionPath string,
+	writeImages bool,
+	writeTimeseries bool,
+	wg *sync.WaitGroup,
+) (common.Callback, common.CallbackTeardown) {
+
+	sess := session.Must(session.NewSession(&aws.Config{Region: aws.String(awsS3Region)}))
+	uploader := s3manager.NewUploader(sess)
+
+	if awsDescriptionPath != "" {
+		awsDescription, err := os.Open(awsDescriptionPath)
+		if err != nil {
+			log.Fatalf("Could not find %v: %v", awsDescriptionPath, err)
+		}
+		wg.Add(1)
+		go func() {
+			key := fmt.Sprintf("ABOUT%v", filepath.Ext(awsDescriptionPath))
+			if project != "" {
+				key = fmt.Sprintf("%v/%v", project, key)
+			}
+			upload(uploader, key, awsDescription)
+			wg.Done()
+		}()
+	}
+
+	callback := func(pkg common.DataRecord) {
+		if pkg.Error != nil {
+			log.Println(pkg.Error)
+			return
+		}
+		if writeImages {
+			switch pkg.Data.(type) {
+			case aez.CCDImage:
+				ccdImage, ok := pkg.Data.(aez.CCDImage)
+				if !ok {
+					log.Print("Could not understand packet as CCDImage, this should be impossible.")
+					break
+				}
+				img, imgFileName := ccdImage.Image(pkg.Buffer, project, pkg.Origin.Name)
+
+				wg.Add(1)
+				go func() {
+					pngBuffer := bytes.NewBuffer([]byte{})
+					png.Encode(pngBuffer, img)
+					upload(uploader, imgFileName, pngBuffer)

```

Comment:
The aws_s3_writer is very similar to disk_writer is it possible to generalize the control flow of saving to storage? upload() or "save to file" seems to be the discrimintor between the two callbacks?

Reply:
I was considering this, but I don't think it's the proper time for it. There's an inherent risk of over engineering it and at least the time-series data should also have been handled by the aws writer.